src/librustc_trait_selection/traits/error_reporting/mod.rs

```diff
@@ -1744,6 +1804,44 @@ impl<'a, 'tcx> InferCtxtPrivExt<'tcx> for InferCtxt<'a, 'tcx> {
     }
 }
 
+/// Look for type `param` in an ADT being used only through a reference to confirm that suggesting
+/// `param: ?Sized` would be a valid constraint.
+struct FindTypeParam {
+    param: rustc_span::Symbol,
+    invalid_spans: Vec<Span>,
+    nested: bool,
+}
+
+impl<'v> Visitor<'v> for FindTypeParam {
+    type Map = rustc_hir::intravisit::ErasedMap<'v>;
+
+    fn nested_visit_map(&mut self) -> hir::intravisit::NestedVisitorMap<Self::Map> {
+        hir::intravisit::NestedVisitorMap::None
+    }
+
+    fn visit_ty(&mut self, ty: &hir::Ty<'_>) {
```

Comment:
So, this analysis doesn't seem right. For example, consider an example like this:

```rust
struct Foo<T> {
    x: Vec<T>, // requires `T: Sized`
    y: Box<T>, // does not require `T: Sized`
}
```

if I'm reading the code correctly, it is going to highlight both of those uses of `T`?

Doing this correctly is going to be a *touch* tricky at least, but we could make it more precise in various ways. For example, we could go over the list of fields and look at their `Ty<'tcx>` type to see if its WF conditions require that `T: Sized`, and only then walk over the HIR for the type.

Reply:
We now only point at the uses of `T` when it is bare, otherwise we _assume_ that they allow `?Sized`. That example will suggest `T: ?Sized`:

```
error[E0277]: the size for values of type `T` cannot be known at compilation time
 --> file6.rs:6:21
  |
1 | struct Foo<T> {
  |            - required by this bound in `Foo`
...
6 | struct X<T: ?Sized>(Foo<T>);
  |          -          ^^^^^^ doesn't have a size known at compile-time
  |          |
  |          this type parameter needs to be `std::marker::Sized`
  |
  = help: the trait `std::marker::Sized` is not implemented for `T`
  = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
help: consider relaxing the implicit `Sized` restriction
  |
1 | struct Foo<T: ?Sized> {
  |             ^^^^^^^^
```

When applying it you get

```
error[E0277]: the size for values of type `T` cannot be known at compilation time
   --> file6.rs:2:5
    |: aborting due to previous error
1   | struct Foo<T: ?Sized> {
    |            - this type parameter needs to be `std::marker::Sized`
2   |     x: Vec<T>, // requires `T: Sized` 60☶)$ vim file6.rs
    |     ^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `T`
    = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
```

I am indeed not verifying whether `Vec` or `Box` have a `T: ?Sized` or `T: Sized` constraint. That would be a good thing to add, but I feel that the current behavior is _good enough_ before adding more code to probe `Vec` and `Box` to identify what their bounds are.

Reply:
OK -- can you maybe add some comments into the code explaining that a bit? i.e., what is the goal of this visitor and can you clarify where it makes 'safe assumptions'?

Reply:
Added a comment.